### ISSUE TITLE
Multi-tagging ?

### ISSUE THREAD

**Issue by leomignot** [emilienschultz/activetigger#124]
Do you need or would it be useful to have the ability to apply multiple tags to a text in the annotation process ? I suppose it's not the usual workflow and it could be a pain for training. So, I don't really believe it, but I'm raising the point just in case.

**Comment by jboelaert** [comment]
That is a feature I've been wanting to implement for years, but indeed it
implies a very different approach for much of the workflow. For instance,
how do you define the most ambiguous element to tag in Active mode when you
have multiple classifiers?
I think it's one of the things to put on the wishlist for a (distant) v2.

**Comment by emilienschultz** [comment]
After the presentation in Grenoble, exchanges with Frédérique Gonthier (a need for annotating Gilets Jaunes's dataset and debrief with @eollion , it may be a path:

- 2 types of scheme : multiclass / multilabel

And in the multilabel :

- a specific component allowing to select the labels in a list / recorded as a list/json in the database (rather than clickable labels)
- the simplemodel is always calculated as a binary classifier on one specific label (no active "in general", but specific for a label), need to select a label
- same for the bert fine-tuning, on binomial classification, need to select a label
- disable the possibility to modify annotation in the explore mode
- modify the way to compute statistics to have broader statistics
- modify the export

It could be an important feature to implement in the next weeks

**Comment by emilienschultz** [comment]
I did a first version of multilabel. Yeah !

**Comment by jboelaert** [comment]
Wow, trop bien !

Une remarque sur le Bert : on peut faire du multiclasse directement, plutôt
que sur chaque classe.
Ici une proposition de Llama, pas testé mais ça a l'air legit.

If you want the model to classify a text into several categories (i.e., a
multi-label classification problem), the approach is slightly different.
The key difference is in how the classification head is defined and how the
loss is calculated.

Here's an example of how you can adapt the previous code for a multi-label
classification task:

```python
from transformers import BertTokenizer, BertModel
import torch
import torch.nn as nn
import torch.nn.functional as F

# Define the number of classes for your multi-label problem
num_classes = 8  # Example: 8 classes

# Load pre-trained BERT tokenizer and model
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertModel.from_pretrained('bert-base-uncased')

# Define a custom classification head for multi-label classification
class MultiLabelClassificationHead(nn.Module):
    def __init__(self, num_classes):
        super(MultiLabelClassificationHead, self).__init__()
        self.dropout = nn.Dropout(0.1)
        self.classifier = nn.Linear(bert_model.config.hidden_size,
num_classes)

    def forward(self, x):
        x = self.dropout(x)
        x = self.classifier(x)
        return x

# Initialize the classification head
classification_head = MultiLabelClassificationHead(num_classes)

# Your data preparation goes here...
# For demonstration purposes, assume 'text_samples' is a list of strings
you want to classify
text_samples = ["This is a sample text.", "Another sample for
classification."]

# Preprocess the text samples
inputs = tokenizer(text_samples, return_tensors="pt", padding=True,
truncation=True)

# Forward pass through BERT
outputs = bert_model(**inputs)

# Get the last hidden state of the sequence
last_hidden_state = outputs.last_hidden_state

# Get the embedding of the [CLS] token, which represents the entire sequence
sequence_embeddings = last_hidden_state[:, 0, :]

# Forward pass through the classification head
logits = classification_head(sequence_embeddings)

# Apply sigmoid to get probabilities (since it's a multi-label problem)
probabilities = torch.sigmoid(logits)

print(probabilities)
```

In this example:
- We define a custom classification head `MultiLabelClassificationHead`
that takes the output of the BERT model and outputs a vector of logits for
each class.
- We use the `BertModel` class instead of `BertForSequenceClassification`,
since we're defining our own classification head.
- We apply the sigmoid function to the logits to get probabilities, since
it's a multi-label problem and we want to predict the probability of each
class independently.

Note that in a multi-label problem, you would typically use a different
loss function, such as binary cross-entropy loss, to train the model. You
would also need to modify the training loop to accommodate the multi-label
setup.

Here's an example of how you could define the loss function and train the
model:

```python
# Define the loss function
criterion = nn.BCEWithLogitsLoss()

# Define the optimizer
optimizer = torch.optim.Adam(list(bert_model.parameters()) +
list(classification_head.parameters()), lr=1e-5)

# Train the model
for epoch in range(5):
    optimizer.zero_grad()
    outputs = bert_model(**inputs)
    last_hidden_state = outputs.last_hidden_state
    sequence_embeddings = last_hidden_state[:, 0, :]
    logits = classification_head(sequence_embeddings)
    loss = criterion(logits, labels)  # assume 'labels' is a tensor of
shape (batch_size, num_classes)
    loss.backward()
    optimizer.step()
    print(f'Epoch {epoch+1}, Loss: {loss.item()}')
```

In this example, we define the loss function as binary cross-entropy loss
with logits, and we train the model using the Adam optimizer. We assume
that `labels` is a tensor of shape `(batch_size, num_classes)`, where each
element is a binary label indicating whether the corresponding class is
present or not.
